Update updateruntimeassemblies.ps1 to exclude  assemblies with resolution policy private

pragnagopa opened the pull request:
<!-- Please provide all the information below.  -->

### Issue describing the changes in this PR

- Moved existing `runtimeassemblies.txt` to test project.
- Added `src/runtimeassemblies.txt` to `.gitginore`
- Added test to verify generated and existing list of assemblies.

Related issue https://github.com/Azure/azure-functions-vs-build-sdk/issues/422

### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [x] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [ ] I have added all required tests (Unit tests, E2E tests)

<!-- Optional: delete if not applicable  -->
### Additional information

Additional PR information
Generated runtimeassemblies.txt
[runtimeassemblies.txt](https://github.com/Azure/azure-functions-host/files/5052810/runtimeassemblies.txt)

```
Added: Microsoft.ApplicationInsights.dll
Added: Microsoft.Azure.WebJobs.Script.Abstractions.dll
Removed: Microsoft.AspNetCore.Authentication.Facebook.dll
Removed: Microsoft.AspNetCore.Authentication.Google.dll
Removed: Microsoft.AspNetCore.Authentication.MicrosoftAccount.dll
Removed: Microsoft.AspNetCore.Authentication.OpenIdConnect.dll
Removed: Microsoft.AspNetCore.Authentication.Twitter.dll
Removed: Microsoft.AspNetCore.Authentication.WsFederation.dll
Removed: Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.dll
Removed: Microsoft.AspNetCore.Identity.UI.dll
Removed: Microsoft.AspNetCore.Identity.UI.Views.dll
Removed: Microsoft.AspNetCore.MiddlewareAnalysis.dll
Removed: Microsoft.AspNetCore.NodeServices.dll
Removed: Microsoft.AspNetCore.Owin.dll
Removed: Microsoft.AspNetCore.Server.Kestrel.Https.dll
Removed: Microsoft.AspNetCore.Server.Kestrel.Transport.Abstractions.dll
Removed: Microsoft.AspNetCore.SpaServices.dll
Removed: Microsoft.AspNetCore.SpaServices.Extensions.dll
Removed: Microsoft.Data.Edm.dll
Removed: Microsoft.Data.OData.dll
Removed: Microsoft.Extensions.Caching.SqlServer.dll
Removed: Microsoft.IdentityModel.Protocols.WsFederation.dll
Removed: Microsoft.IdentityModel.Tokens.Saml.dll
Removed: Microsoft.IdentityModel.Xml.dll
Removed: SOS.NETCore.dll
Removed: System.Net.WebSockets.WebSocketProtocol.dll
Removed: System.Spatial.dll

```

pragnagopa commented:
@brettsam / @fabiocav - Updates in this PR are generated by running script : https://github.com/Azure/azure-functions-host/blob/dev/tools/ExtensionsMetadataGenerator/src/ExtensionsMetadataGenerator/updateruntimeassemblies.ps1

Should there be tracking item to publish Microsoft.Net.Sdk.Functions consuming latest version of ExtensionsMetadataGenerator package?

brettsam commented on tools/ExtensionsMetadataGenerator/src/ExtensionsMetadataGenerator/runtimeassemblies.txt:
```diff
@@ -1,17 +1,24 @@
+Microsoft.Azure.Cosmos.Table.dll
+Microsoft.Azure.Storage.Blob.dll
+Microsoft.Azure.Storage.File.dll
+Microsoft.Azure.Storage.Queue.dll
```
This makes me nervous -- a lot of people use newer versions of these Storage assemblies and we can't just remove them if they're there (same with App Insights below).

pragnagopa replied:
the reason ApplicationIsights is added here is  - it is listed in runtimeassemblies.json - https://github.com/Azure/azure-functions-host/blob/dev/src/WebJobs.Script/runtimeassemblies.json#L53-L56 and we are enforcing the version loaded by host and not what users brings in. 

Please do correct me if I misunderstood. If these assemblies should not be cleaned up then we should remove them from runtimeassemblies.json ?

fabiocav reviewed (dismissed):
As discussed, this file is automatically generated at build time. A change is not required.

brettsam commented:
No but we should modify the script to accept exclusions from the list. I think all of the Microsoft.Azure.Storage and Cosmos ones need to be excluded or we'll have a bunch of people run into this immediately.

pragnagopa commented:
> No but we should modify the script to accept exclusions from the list

should they be marked as `private` instead of `runtimeVersion` ?

fabiocav commented:
We can update the generation script.
@pragnagopa the storage ones Brett is referring to should already be marked as private. Is that not the case?

fabiocav commented:
I'm just noticing that the script here needs to be updated to ignore private assemblies:
https://github.com/Azure/azure-functions-host/blob/dev/tools/ExtensionsMetadataGenerator/src/ExtensionsMetadataGenerator/updateruntimeassemblies.ps1

pragnagopa commented:
@fabiocav 

>  the storage ones Brett is referring to should already be marked as private. Is that not the case?

You are right. Storage assemblies are marked as private but `Microsoft.ApplicationInsights` is not. Should this marked as private as well?

Looking into script to exclude assemblies marked as private.

fabiocav commented:
No. That caused problems in the past.

pragnagopa committed:
Update script to exclude private assemblies

pragnagopa commented:
@fabiocav  / @brettsam - Updated the PR with udpates to script. Please take a look at the generated runtimeassemblies.txt that is attached to this PR.

pragnagopa commented:
/azp run

brettsam commented:
Yeah sorry I was wrong about the App Insights assembly... it's okay if that's removed as we always return the one used by the host.

pragnagopa commented:
@fabiocav - should I delete checked in file: https://github.com/Azure/azure-functions-host/commits/dev/tools/ExtensionsMetadataGenerator/src/ExtensionsMetadataGenerator/runtimeassemblies.txt as the script generates it?

pragnagopa commented:
Ping for review!

brettsam commented:
I do think we should delete that file as it's pretty misleading and not what necessarily gets built in. However I also don't like that we wouldn't have any artifact to compare between versions to see if something changed. What if...

1. We don't run the script during the build but instead have a *test* that runs the script and compares the output to the checked-in file. If there is a difference, the test writes a clear exception message like "A change was detected in the runtime assemblies, please run the ps1 script and submit a PR to check in the change..."
2. The build actually uses the checked-in file without generating it.

Yes, it's a little more work but I think it's more clear and we get automatically alerted when there's a change to this file that requires a release.

pragnagopa commented:
Thanks @brettsam . I will add a test.

fabiocav commented:
@pragnagopa to clarify, are you going to add a test and change the flow? The process is pretty safe now as there's no way to have a build without an update, so the test would need to ensure that happens.

fabiocav commented on tools/ExtensionsMetadataGenerator/src/ExtensionsMetadataGenerator/runtimeassemblies.txt:
```diff
@@ -1,357 +0,0 @@
-Microsoft.AspNetCore.dll
```
I don't think we should delete this file. Having a change history for this would be helpful.

pragnagopa replied:
If only generated file that is part of the buildoutput is needed, then including generated file as embedded resource when building nuget package should work right? We can have checked in version of this file in tests?

fabiocav replied:
We could, this also serves to keep the change history here if you intend to update the process, since in order to pass, the test would need this file to be updated.

pragnagopa commented:
Moved existing file to test project. For this time, changes in PR should serve as verification of the diff.  Added test that fails if generated list changes.

Please review!

pragnagopa commented:
/azp run

pragnagopa commented:
/azp run

fabiocav commented:
@pragnagopa can you resubmit a review request when this is ready for review?

pragnagopa commented:
@fabiocav / @brettsam  - PR is ready for review. PR build is failing with Nuget errors. I am follow up separately on  Nuget errors.

pragnagopa commented:
/azp run

pragnagopa commented:
Nuget issues resolved. Please review!

pragnagopa commented:
Ping! to bump for reviews! Thanks!

brettsam commented on tools/ExtensionsMetadataGenerator/src/ExtensionsMetadataGenerator/ExtensionsMetadataGenerator.csproj:
```diff
@@ -1,4 +1,4 @@
-﻿<Project Sdk="Microsoft.NET.Sdk">
+﻿<Project Sdk="Microsoft.NET.Sdk" InitialTargets="UpdateRuntimeAssemblies">
```
Is this required? Doesn't the `BeforeTargets="Build"` handle this?

pragnagopa replied:
If `runtimeassemblies.txt` file does not exist, build fails without `InitialTargets` . Let me know if I missed anything.

brettsam commented on tools/ExtensionsMetadataGenerator/test/ExtensionsMetadataGeneratorTests/RuntimeAssembliesUpdateTests.cs:
```diff
@@ -0,0 +1,33 @@
+            result = result.Union(diffRemoved.Select(s => $"Removed: {s}"));
+
+            File.WriteAllLines("Result.txt", result);
+            Assert.False(result.Any(), $"Generated runtimeassemblies.txt does not match existing list. Please look at {DiffListFileName} for changes. Verify changes and update contents in {ExistingRuntimeAssembliesFileName}");
```
Can you have this just list them in the failure text rather than write to a file? That way if someone hits this on a CI machine it's clear what they hit.

brettsam replied:
Also maybe put the full path to runtimeassemblies.txt and the Existing one just so it's clear where it is?

brettsam reviewed (commented):
couple small comments....

pragnagopa committed:
Addressing CR

pragnagopa committed:
fix file names
